#3312 | Large Pulse backend mock

IceKhan13 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Create a large (100Q) Pulse backend mock and/or create a util method or a script to generate reasonable Pulse-enabled backend

Issue https://github.com/Qiskit/qiskit-terra/issues/3312


### Details and comments
The pulse module of Qiskit has yet to be benchmarked. The purpose of benchmarking is to provide a deterministic way to assess performance of software over time. There is room for improvement in the performance of pulse operations; building benchmarks will allow us to quantify progress and justify code refactors.

lcapelluto commented on qiskit/test/mock/backends/fake_100q/conf_100q.json:
```diff
@@ -0,0 +1 @@
+{"qubit_lo_range":[],"memory":true,"rep_times":[1000],"gates":[],"backend_version":"1.2.4","coupling_map":[[0, 1]],"description":"100 qubit device pulse","max_shots":8192,"n_uchannels":46,"acquisition_latency":[],"max_experiments":900,"conditional_latency":[],"n_registers":1,"dt":2.2222222222222221e-10,"meas_lo_range":[],"basis_gates":["u1","u2","u3","cx","id"],"backend_name":"alt_almaden","u_cha ...
```
I see these are still just getting started, but would you mind reformatting these 3 files? Not sure what editor you use, but I've used this in the past with Sublime: https://packagecontrol.io/packages/Pretty%20JSON I'm sure there are similar solutions regardless of your env :)

lcapelluto commented on qiskit/test/mock/backends/fake_100q/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
```
This backend also has to be added to `qiskit/test/mock/backends/__init__.py`

taalexander commented on qiskit/test/mock/backends/__init__.py:
```diff
@@ -31,3 +31,4 @@
 from .johannesburg import FakeJohannesburg
 from .boeblingen import FakeBoeblingen
 from .rochester import FakeRochester
+from .fake_100q import FakeOpenPulse100Q
```
Perhaps this should live in `test.mock` since `backends` currently seems reserved for a physical backend?

taalexander commented on qiskit/test/mock/backends/fake_100q/conf_100q.json:
```diff
@@ -0,0 +1,68 @@
+        2
+    ],
+    "meas_lo_range": [],
+    "meas_map": [
```
There must be an entry for every qubit, eg., every qubit index must live in the meas_map.

taalexander commented on qiskit/test/mock/backends/fake_100q/conf_100q.json:
```diff
@@ -0,0 +1,68 @@
+        ]
+    ],
+    "memory": false,
+    "n_qubits": 10,
```
```suggestion
    "n_qubits": 100,
```

taalexander commented on qiskit/test/mock/backends/fake_100q/defs_100q.json:
```diff
@@ -0,0 +1,592 @@
+{
+    "buffer": 10,
+    "cmd_def": [
```
We require a full basis gate set for all qubits and CX gates to connect them. These could be copied from a preexisting backend.

taalexander commented on qiskit/test/mock/backends/fake_100q/defs_100q.json:
```diff
@@ -0,0 +1,592 @@
+    ],
+    "converter": "",
+    "instruction_schedule_map": "",
+    "meas_freq_est": [
```
The number of entries should be equal to the number of qubits.

taalexander commented on qiskit/test/mock/backends/fake_100q/defs_100q.json:
```diff
@@ -0,0 +1,592 @@
+        6500000000.0,
+        6500000000.0
+    ],
+    "pulse_library": [],
```
We require pulses to be present for all entries in the `cmd_def`.

taalexander commented on qiskit/test/mock/backends/fake_100q/props_100q.json:
```diff
@@ -0,0 +1,1230 @@
+{
+    "backend_name": "100q",
```
There are also only entries here for 10 qubits.

taalexander commented on qiskit/test/mock/backends/fake_100q/fake_100q.py:
```diff
@@ -0,0 +1,265 @@
+
+
+# TODO: move to utils or remove if not necessary
+class FakeBackendBuilder(object):
```
I think you are on the right track with this.

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -13,93 +13,86 @@
     """
 
+    _SINGLE_QUBIT_GATES = ['id', 'u1', 'u2', 'u3']
+    _DEFAULT_BASIS_GATES = ['id', 'u1', 'u2', 'u3', 'cx']
```
This could just be in the main body of init, the standard pattern is
```python
if basis_gates is None:
    basis_gates = ['id', 'u1', 'u2', 'u3', 'cx']
```
and then document the default in the docstring.

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -13,93 +13,86 @@
 
+    _SINGLE_QUBIT_GATES = ['id', 'u1', 'u2', 'u3']
+    _DEFAULT_BASIS_GATES = ['id', 'u1', 'u2', 'u3', 'cx']
+    _DEFAULT_T1 = 113.3
```
These could be directly in keyword argument.

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -13,93 +13,86 @@
+                 ):
         """
 
         Args:
```
Could you please document the parameters here and how/what be set to use.

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -0,0 +1,308 @@
+# -*- coding: utf-8 -*-
```
Do you think there might be a better name for this module? Maybe fake_backend_builder?

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -0,0 +1,308 @@
+                           qubit_frequency: Optional[float] = None,
+                           qubit_readout_error: Optional[float] = None) -> Type[FakeBackend]:
+    """
+    Example:
```
Could you please document the args in more detail?

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -0,0 +1,308 @@
+                      default=lambda o: '')
+
+
+def fake_backend_generator(name: str,
```
this would probably be better as `generate_fake_backend` which is a verb while still indicating what it returns.

IceKhan13 replied:
placed it withing `FakeBackendBuilder` as `build` method for now

taalexander commented on qiskit/test/mock/utils/backend.py:
```diff
@@ -0,0 +1,308 @@
+                      default=lambda o: '')
+
+
+def fake_backend_generator(name: str,
```
I think something is wrong here, perhaps this function isn't needed and `FakeBackendBuilder` could become `FakeBackend` which you could then instantiate with the desired parameters?

IceKhan13 commented:
Team, I need some input from you 😄 
I'm trying to write tests for generated by builder backends (create circuit and run it on the backend.)

It goes as usual:
```python
desired_vector = [1 / math.sqrt(2), 0, 0, 1 / math.sqrt(2)]
qr = QuantumRegister(2, "qr")
qc = QuantumCircuit(qr)
qc.initialize(desired_vector, [qr[0], qr[1]])
```
then I build a backend and try to transpile cirquit, create schedule and assemble qobj:
```python
FakeOpenPulse10Q = FakeBackendBuilder("Tashkent", n_qubits=10).build()
backend = FakeOpenPulse10Q()

experiments = transpile(qc, backend=backend)
experiments = schedule(circuits=experiments, backend=backend)
qobj = assemble(experiments, backend=backend)
```
and everything works but to finish logical sequence for `execute` I need to run `backend.run(qobj)`.
And here execution fails due to pickle error:
```
_pickle.PicklingError: Can't pickle <class 'abc.FakeOpenPulse10Q'>: attribute lookup FakeOpenPulse10Q on abc failed
```
Same thing happens when I try to use `FakeOpenPulse2Q` (existing) backend. But in case of existing backend it failed due to inaability to pickle local object:
```
AttributeError: Can't pickle local object 'QobjToInstructionConverter.convert_frame_change.<locals>.gen_fc_sched'
```
This can be solved (I assume) by using `__setstate__` and `__getstate__` to manually pickle-unpickle objects. 

Anyhow both of this errors occurs in `run` function, where in `FakeBackend` we are using `FakeJob` which uses `futures.ProcessPoolExecutor()`. And after 15 seconds of googling I noticed that dynamic classes has issue with pickling (which is happening while using process pools).

I think, I can create other `FakeJob` which will not use pools and use it in generated fake backends `run` method. 🤔 Maybe it's 2am and I need to sleep a little to came up with better solution 😄 

Any suggestions how to overcome this?) 


P.S
Is is ok that I always mention your names in PR discussions @taalexander @lcapelluto ? It probably creates a lot of notifications in your accounts. Sorry for that. I can just leave a comments and wait 😄

taalexander commented:
It's totally ok to ping me 😄. I was tried to run the sample code 

```python
import math

from qiskit import QuantumRegister, QuantumCircuit, transpile, schedule, assemble
from qiskit.test.mock.utils.fake_backend_builder import FakeBackendBuilder

desired_vector = [1 / math.sqrt(2), 0, 0, 1 / math.sqrt(2)]
qr = QuantumRegister(2, "qr")
qc = QuantumCircuit(qr)
qc.initialize(desired_vector, [qr[0], qr[1]])

FakeOpenPulse10Q = FakeBackendBuilder("Tashkent", n_qubits=10).build()
backend = FakeOpenPulse10Q()
experiments = transpile(qc, backend=backend)
experiments = schedule(circuits=experiments, backend=backend)
qobj = assemble(experiments, backend=backend)
```
but am getting the error

``TranspilerError: "The 'layout' must be full (with ancilla)."``

If I look at the coupling_map, it includes too many qubits.
```bash
>>> backend.configuration().coupling_map
[[0, 4], [1, 2], [2, 3], [2, 6], [3, 4], [4, 8], [5, 6], [6, 7], [6, 10], [7, 8], [8, 12], [9, 10]]
```

taalexander commented:
The backend is missing a parameter `rep_times` I think.

taalexander commented:
I can, however, recreate your error with 

```python
import math

from qiskit import QuantumRegister, QuantumCircuit, transpile, schedule, assemble
from qiskit.test.mock.utils.fake_backend_builder import FakeBackendBuilder
from qiskit.test.mock import FakeOpenPulse2Q

desired_vector = [1 / math.sqrt(2), 0, 0, 1 / math.sqrt(2)]
qr = QuantumRegister(2, "qr")
qc = QuantumCircuit(qr)
qc.initialize(desired_vector, [qr[0], qr[1]])

FakeOpenPulse10Q = FakeBackendBuilder("Tashkent", n_qubits=4, coupling_map=[[0, 1], [1,2], [2,3]]).build()
#backend = FakeOpenPulse2Q()
backend = FakeOpenPulse10Q()
experiments = transpile(qc, backend=backend)
experiments = schedule(circuits=experiments, backend=backend)
qobj = assemble(experiments, backend=backend, rep_time=100)

job = backend.run(qobj)
result = job.result()
```
It seems like this is trying to pickle the backend object in multiprocessing which may not be possible. I'm not sure why this is necessary perhaps @mtreinish has an idea? Maybe the problem is that `run_job` for the future is a class method?

taalexander commented:
I think this method is broken and untested.

taalexander commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,301 @@
+        def defaults(cls) -> PulseDefaults:
+            return self.build_defaults()
+
+        return type('FakeOpenPulse{}Q'.format(self.n_qubits),
```
I think this is overthinking how to create the backend. Why not
```python
backend = FakeBackend(...)
```
or just have the FakeBackendBuilder return an instance of `FakeBackend`?

IceKhan13 replied:
makes much more sense :) done

taalexander commented:
Did merging in master help?

IceKhan13 commented:
Fake job changed. So, no it's using Aer simulator and reads hamiltonian. Now I need to generate Hamiltonian correctly and continue with tests. But all of your comments above correct and it helped :) 
I'll get back tonight if I'll not be able to run it today.

IceKhan13 commented:
@taalexander I'm trying to run 
```python
        self.backend = FakeBackendBuilder("Tashkent", n_qubits=4).build()
        ...
        desired_vector = [1 / math.sqrt(2), 0, 0, 1 / math.sqrt(2)]
        qr = QuantumRegister(2, "qr")
        cr = ClassicalRegister(2, 'cr')
        qc = QuantumCircuit(qr, cr)
        qc.initialize(desired_vector, [qr[0], qr[1]])
        qc.measure(qr[0], cr[0])
        qc.measure(qr[1], cr[1])

        experiments = transpile(qc, backend=self.backend)
        experiments = schedule(circuits=experiments, backend=self.backend)
        qobj = assemble(experiments, backend=self.backend)

        job = self.backend.run(qobj)
```

which is code from here -> https://github.com/Qiskit/qiskit-terra/pull/3797/commits/7fe4bd9ca91d27c256572ec6b655dcbb38549a44#diff-5b5c1ce1e7ad1fef068081a327804e4aR31

and I'm getting
```
qiskit-terra/lib/python3.7/site-packages/qiskit/providers/aer/pulse/qobj/digest.py", line 323, in build_pulse_arrays
    oplist_to_array(pulse['samples'], pulses, pulses_idx[ind - 1])
TypeError: Argument 'A' has incorrect type (expected list, got numpy.ndarray)
```
which is so funny because that is the only change which is still not included into Aer release 😆  
here is the change which will fix error above https://github.com/Qiskit/qiskit-aer/compare/0.4.1...stable/0.4#diff-efd70aeb09d7cc715321c96acb7a7af0R324

Should I wait till release of Aer?

IceKhan13 commented:
I just talked to Matthew and release of Aer will be within 2 weeks. Meanwhile I'll clean the code up and finish all remaining todos.

taalexander commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,348 @@
+        return self.coupling_map if self.coupling_map else self._generate_cmap()
+
+    def _generate_cmap(self) -> List[List[int]]:
+        """Generate Almaden like coupling map."""
```
I appreciate the work you put into this, but by default, it may be more straightforward and predictable if it is just an all to all coupling_map.

IceKhan13 replied:
All-to-all you mean planar grid-like graph like this?

```
* --- * --- *
 |    |    |
* --- * --- *
```

Or really all to all cmap with `0.5*n*(n-1)` connections ?

taalexander replied:
I initially meant the latter, but I think you're right the former would be better.

taalexander commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,348 @@
+            open_pulse=True,
+            memory=False,
+            max_shots=65536,
+            gates=[GateConfig(name='TODO', parameters=[], qasm_def='TODO')],
```
Perhaps we can just leave this empty for now if they are unavailable.

taalexander commented on test/mock/test_fake_backend_builder.py:
```diff
@@ -0,0 +1,127 @@
+from qiskit.test.mock.utils.fake_backend_builder import FakeBackendBuilder
+
+
+class FakeBackendBuilderTest(QiskitTestCase):
```
```suggestion
class TestFakeBackendBuilder(QiskitTestCase):
```

taalexander commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,348 @@
+            'cmd_def': cmd_def
+        })
+
+    def dump(self, folder: str):
```
Perhaps this method should be named `save` and if it exists maybe there should be a classmethod `load` that loads from a dumped folder?

IceKhan13 commented:
Ok, I've tried to write save and load methods, but faced a problem with `to_dict` methods. Apparently, `to_dict` is not converting all fields of `BackendProperties` correctly therefore `json.dump` raises error `TypeError: keys must be str, int, float, bool or None, not tuple`. I notices there were recent changes in this methods, maybe it's bug somewhere, because I remember it was working before I updated master.

I removed those methods for now and probably we need to create an issue for `BackendProperties` `to_dict` method. 

I'll leave it here for now :) 
```python
    def save(self, folder: str):
        """Saves backend configuration files to specified folder."""
        with open('{0}/props.json'.format(folder), 'w') as file:
            json.dump(self.build_props().to_dict(), file, indent=4, sort_keys=True)

        with open('{0}/conf.json'.format(folder), 'w') as file:
            json.dump(self.build_conf().to_dict(), file, indent=4, sort_keys=True)

        with open('{0}/defs.json'.format(folder), 'w') as file:
            json.dump(self.build_defaults().to_dict(), file,
                      indent=4, sort_keys=True,
                      default=lambda o: '')

    @classmethod
    def load(cls, folder: str):
        """Loads saved backend configuration from specified folder."""
        with open('{0}/props.json'.format(folder), 'r') as file:
            props = BackendProperties.from_dict(json.load(file))

        with open('{0}/conf.json'.format(folder), 'r') as file:
            conf = PulseBackendConfiguration.from_dict(json.load(file))

        with open('{0}/defs.json'.format(folder), 'r') as file:
            defs = PulseDefaults.from_dict(json.load(file))

        backend = FakeBackend(conf)
        backend.defaults = defs
        backend.properties = props
        return backend
```

taalexander commented:
Ok, that makes sense. Thanks for the update 😄

IceKhan13 commented:
Shoot! I was not quick enough removing WIP and Luciano marked this PR as draft 😄 
@taalexander @lcapelluto  Anything else I need to consider here?

mtreinish commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+
+        self.coupling_map = cmap
+
+        return cmap
```
It's a bit weird to have a private method like this both return the coupling map and set the property to it. In this specific case you end up setting `self.coupling_map` twice when it gets called from `cmap()`

mtreinish commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+        return self.coupling_map if self.coupling_map else self._generate_cmap()
+
+    def _generate_cmap(self) -> List[List[int]]:
+        """Generate default grid-like coupling map."""
```
Is there a reason to just not use `CouplingMap.from_grid`? https://qiskit.org/documentation/stubs/qiskit.transpiler.CouplingMap.from_grid.html#qiskit.transpiler.CouplingMap.from_grid

IceKhan13 replied:
I think `from_grid` will not work for 99 qubits. Or it's ok to have dangling edges?

mtreinish commented on releasenotes/notes/large-mock-backend-builder-af0623b43bdcc35e.yaml:
```diff
@@ -0,0 +1,17 @@
+  - |
+    A new class :py:class:`~qiskit.test.mock.utils.FakeBackendBuilder` has been added
+    which allows users to create configurable mock backends.
+    Example:
```
```suggestion
    Example::
```

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+                 version: Optional[str] = None,
+                 coupling_map: Optional[List[List[int]]] = None,
+                 basis_gates: Optional[List[str]] = None,
+                 qubit_t1: Optional[float] = None,
+                 qubit_t2: Optional[float] = None,
```
It seems like this should accept an array of length `n_qubits`

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+
+        for gate in self.basis_gates:
+            parameters = [Nduv(date=self.now, name='gate_error', unit='', value=1.0),
+                          Nduv(date=self.now, name='gate_length', unit='', value=0.)]
```
This `value` should match the defaults `cmd_def`, for example, `u3` seems to have 4 samples. This is translated to seconds by multiplying 4 by `configuration().dt`. The value in seconds goes here

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+            ])
+
+        for gate in self.basis_gates:
+            parameters = [Nduv(date=self.now, name='gate_error', unit='', value=1.0),
```
`1` might be a problematic gate error if this is ever used by the transpiler. Maybe something like 0.01

IceKhan13 replied:
https://github.com/Qiskit/qiskit-terra/pull/3797/commits/a66464df5a1c1ac8eaff73dff1415cfcf71a9ebb

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+                                      qubits=[qubit1, qubit2],
+                                      parameters=parameters))
+            else:
+                raise QiskitError("{gate} is not supported by fake backend builder."
```
👍

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+
+        qubit_freq_est = np.linspace(4.9, 5.1, self.n_qubits).tolist()
+        meas_freq_est = np.linspace(6.4, 6.6, self.n_qubits).tolist()
+        buffer = 10
```
This has been deprecated. I think it still has to be supplied, but let's just set it to `0`

IceKhan13 replied:
https://github.com/Qiskit/qiskit-terra/pull/3797/commits/a66464df5a1c1ac8eaff73dff1415cfcf71a9ebb

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+
+        cmd_def = [measure_command]
+
+        for i in range(self.n_qubits):
```
I think this should loop over `self.single_qubit_gates` and build a definition for each qubit, for each gate

IceKhan13 replied:
https://github.com/Qiskit/qiskit-terra/pull/3797/commits/a66464df5a1c1ac8eaff73dff1415cfcf71a9ebb

lcapelluto commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+                }).to_dict()
+            ]
+
+        for couple in self.cmap:
+            qubit1, qubit2 = couple
```
```suggestion
        for qubit1, qubit2 in self.cmap:
```

taalexander commented on qiskit/test/mock/utils/fake_backend_builder.py:
```diff
@@ -0,0 +1,331 @@
+        if basis_gates is None:
+            basis_gates = ['id', 'u1', 'u2', 'u3', 'cx']
+
+        if qubit_t1 is None:
+            qubit_t1 = 113.3
+
+        if qubit_t2 is None:
+            qubit_t2 = 150.2
```
I think we should avoid additional complexity here. Especially if it adds randomness (I understand we could seed this). I think we should focus on getting this merged and then we can add additional features alter.

taalexander commented:
@mtreinish  and @lcapelluto have your concerns been addressed by the very patient @IceKhan13  😄?

taalexander commented:
> I actually have one small request. Other backends have this format: `backend = Something()`. I think introducing `FakeBackendBuilder().build()` doesn't really gain us anything, since we can't change parameters after initializing and running `build`, right? What do you think @IceKhan13 , @taalexander ?

Are you suggesting that the builder be made just the backend class that is to be initialized?

IceKhan13 commented:
@lcapelluto @taalexander 

I guess if we want to remove build, we should rename class to something like `ConfigurableBackend` and inherit from `FakeBackend` class like 2q and 3q examples we have.

IceKhan13 commented:
ping :)

lcapelluto commented:
@IceKhan13 I like the `ConfigurableBackend` idea. I think it's more inline with what this accomplishes: as it is, everything is provided at init, and the result object can't be used until build is called to get a backend.

mtreinish commented on test/mock/test_fake_backends.py:
```diff
@@ -0,0 +1,64 @@
+    def setUp(self) -> None:
+        self.backend = ConfigurableBackend("Tashkent", n_qubits=4)
+
+    def test_transpile_schedule_and_assemble(self):
```
There is partial overlap here with: https://github.com/Qiskit/qiskit-terra/blob/master/test/python/providers/test_fake_backends.py#L51 but this is different enough that it's probably fine. It might be a good follow up to expand this or the other to cover both use cases and deduplicate things a bit.

jyu00 commented:
@IceKhan13 Can the name be changed to `ConfigurableFakeBackend`? We may have real configurable backends in the future.

MGLetsoalo commented:
Hello, we had a similar project to this one titled: open pulse mock backend for spin based qubits from qiskit camp Africa 2019, the mock qubits where associated with the 13-C nuclear spins coupled to NV centres in a diamond. Based on the computational ability of carbon atoms, it was my believe that these carbon atoms could be used as fake backends that will participate as reactive groups in drug design to combat the covid19 pandemic. If we treat the electron spin as a quantum bus, we can adjust the states that correspond to the syndrome measurements and thus the 13-C  nuclear spins provide a platform to design the proper chemistry that will help mitigate the syndrome. There is a lot of research on the development of viral nanoparticles as a template for organising materials at the nano scales and I think that it is time to consolidate these efforts.

taalexander commented:
I think it is fine to still have an example backend for an NV center system. Would it be possible to use @IceKhan13's work for this and choose the right parameters? Maybe some slight modifications would be required?

MGLetsoalo commented:
Yes, with 100 qubits we can realise a 100*100 tabular qubit ordering, how do I become part of this discussion so that I may be able to present my attempts?

taalexander commented:
If you open up a draft PR with your attempts we can comment directly on it 😄
